### PR TITLE
Inferring and clipping distance and goal occasion for already manipulated containers

### PR DESCRIPTION
**Pull request by cpollok**
Opening and closing designators don't need a distance necessarily anymore. When it's not given a default distance is inferred from the heuristics module. 
In both cases (distance given or not) the distance value is then clipped to fit within the limits of the joint of the container.

Additionally a goal can be given to the action designator consisting of the container and the distance.
It can be used as follows:

```lisp
(let ((?obj (an object
                         (type drawer)
                         (part-of kitchen)
                         (urdf-name sink_area_left_upper_drawer_main)))
               (?dist 0.1))
           (pr2-proj:with-projected-robot
               (exe:perform (an action
                                (goal (container-state ?obj ?dist))
                                (type opening)
                                (arm left)
                                (distance ?dist)
                                (object ?obj)))))
```

### CODE REVIEW

File: cram_3d_world/cram_urdf_environment_manipulation/src/environment-occasions.lisp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+;;; SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+;;; CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+;;; ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+;;; POSSIBILITY OF SUCH DAMAGE.
+
+(in-package :env-man)
+
+(prolog:def-fact-group environment-occasions ()
+  (prolog:<- (:container-state ?container-designator ?distance)
```
Comment (gaya-) [inline cram_3d_world/cram_urdf_environment_manipulation/src/environment-occasions.lisp]: The ``prolog`` here looks bad, just get rid of it, it's being ``use``d.

File: cram_3d_world/cram_urdf_environment_manipulation/src/environment-occasions.lisp [hunk 2]
```diff
@@ -0,0 +1,40 @@
+;;;
+;;; Copyright (c) 2018, Christopher Pollok <cpollok@uni-bremen.de>
```
Comment (gaya-) [inline cram_3d_world/cram_urdf_environment_manipulation/src/environment-occasions.lisp]: 2019

File: cram_common/cram_manipulation_interfaces/src/manipulation-interfaces.lisp [hunk 3]
```diff
@@ -65,3 +65,9 @@ make CRAM_DESIGNATORS package depend on CRAM_MANIPULATION_INTERFACES,
 and man-int is way too high level to make it a dependency of CRAM_CORE.
 This hijacking is kind of an ugly hack that Gaya feels bad about :(."
   (get-location-poses desig))
+
+(defgeneric get-container-opening-distance (container-designator)
```
Comment (gaya-) [inline cram_common/cram_manipulation_interfaces/src/manipulation-interfaces.lisp]: The parameter should be ``container-type'' not a whole designator, such that we can use ``eql'' method overloading stuff.

File: cram_common/cram_object_knowledge/src/environment.lisp [hunk 4]
```diff
@@ -42,7 +42,17 @@
...
 
-;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
 (defmethod man-int:get-action-gripper-opening :heuristics 20 ((object-type (eql :container)))
   0.10)
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defmethod man-int:get-container-opening-distance :heuristics 20 (container-designator)
```
Comment (gaya-) [inline cram_common/cram_object_knowledge/src/environment.lisp]: Should define one for ``:container-prismatic`` and one for ``:container-revolute``.
And the rest should be inferred using ``most-specific`` stuff.
E.g., you call ``(get-container-opening-distance :fridge)`` and get the stuff for ``:container-revolute``.

File: cram_3d_world/cram_urdf_environment_manipulation/src/action-designators.lisp [hunk 5]
```diff
@@ -59,7 +59,14 @@ NAME in the environment BTR-ENVIRONMENT."
     (-> (spec:property ?action-designator (:arm ?arm))
         (true)
         (man-int:robot-free-hand ?_ ?arm))
-    (spec:property ?action-designator (:distance ?distance))
```
Comment (gaya-) [inline cram_3d_world/cram_urdf_environment_manipulation/src/action-designators.lisp]: This is starting to be very copy pasty. Maybe you can look into merging the two designators.

**Comment by gaya-** [comment]
There are 3 new ``defmethods`` in ``object_knowledge/environment.lisp``. When you get the merge conflict, simply keep both your ``defmethod`` and the new ``get-action-grasps`` stuff.

File: cram_3d_world/cram_urdf_environment_manipulation/src/action-designators.lisp (hunk 5 above)
Comment (cpollok) [reply cram_3d_world/cram_urdf_environment_manipulation/src/action-designators.lisp]: I will. But I think this would be better in a separate PR.

**Comment by cpollok** [comment]
> There are 3 new `defmethods` in `object_knowledge/environment.lisp`. When you get the merge conflict, simply keep both your `defmethod` and the new `get-action-grasps` stuff.

So should I merge this one, after the other changes have been made of course?

**Comment by gaya-** [comment]
Closing action without DISTANCE crashes (tries to close to distance 0.4).
Also, SEALING action doesn't do anything anymore, thus, this PR introduces a regression, so I cannot merge this.

**Comment by gaya-** [comment]
Please pull from https://github.com/cram2/cram/tree/cpollok-env-man-distance-revision and fix the above-mentioned issues.

**Comment by gaya-** [comment]
Merged into boxy-melodic.

**Pull request closed without merging**